### PR TITLE
Track and format chat history

### LINKED ISSUE #4: Track and format chat history
The chat format consists in multi-turn conversations between the user and the large language model. Each turn is associated with a role, either `system`, `user` or `assistant`. A chat history can thus be represented as an iterable of tuples:

```python
history = [
    ("system", "This is a system message"),
    ("user", "A user asking a question to the model"),
    ("assistant", "A model responding to the user")
]
```

We need a function which, given a model, will return the corresponding prompt in the correct format. For instance, for Zephyr the corresponding prompt should be:

```ascii
<|system|>
This is a system message</s> 
<|user|>
A user asking a question to the model</s> 
<|assistant|>
A model responding to the user.
```

The simplest way to go about this is to define a function which, given a model name and a sequence of interactions returns the corresponding prompt.

```python
from prompts import to_chat_format

history = [
    ("system", "This is a system message"),
    ("user", "A user asking a question to the model"),
]
prompt = to_chat_format("HuggingFaceH4/zephyr-7b-beta", history)
print(prompt)
# <|system|>
# This is a system message</s> 
# <|user|>
# A user asking a question to the model</s> 
```

However, when we want to pass the prompt to the model for generation we need to append the `<|assistant|>` token. For this we can define a new function, `to_chat_prompt`:

```python
from prompts import to_chat_prompt

history = [
    ("system", "This is a system message"),
    ("user", "A user asking a question to the model"),
]
prompt = to_chat_prompt("HuggingFaceH4/zephyr-7b-beta", history)
print(prompt)
# <|system|>
# This is a system message</s> 
# <|user|>
# A user asking a question to the model</s> 
# <|assistant|>
```

This system is still cumbersome, since we need to append `</s>` to the response, and need to call `to_chat_format` with the model name before every call to the model. Could we simplify this?

```python
class Chat:
    history: List[Tuple[str, str]]

    def __init__(self, model_name: str, system: str=None, tools=None, documents=None):
        raise NotImplementedError

    def __str__(self):
        """Render the prompt corresponding to the history.

        If we want to be able to use the prompt with any library this should return the prompt with the "assistant" keyword?

        As with `Template` we need to
be able to specify the model we are working with.

        """
        raise NotImplementedError

    def __getitem__(self, i):
        return self.history[i]

    def user(self, string: str):
        """Add user input to history"""
        self.history.append(("user", string))

    def assistant(self, string: str):
        """Add user input to history"""
        self.history.append(("user", string))
```

So a session would look like:

```python
from outlines import generate, models


model = models.transformers("HuggingFaceH4/zephyr-7b-beta")

session = Chat("HuggingFaceH4/zephyr-7b-beta", "This is a system message")
session.user("A user asking a question to the model")

result = generate.text(model)(session)
session.assistant(result)
```

The obvious downside of this design is the necessity to specify the model name twice.

**`str(chat)` must return the correctly formatted templates to be able to interface with any library**

This issue (and arguably the repository) was triggered by [this issue](https://github.com/outlines-dev/outlines/issues/756) in Outlines and inspired by [this repository](https://github.com/chrsbats/outlines-chat) opened by a member of the Outlines community.

### PR DESCRIPTION
**Pull request by rlouf**
Given the multiplicity of formats, formatting the prompt for chat workflows with open models can be a real hassle and is error-prone. In this PR we introduce a `Chat` class that allows users to track the conversation and easily print the corresponding prompt.

Closes #4 

We can initialize a chat session with an optional system prompt:

```python 
from prompts import Chat

session = Chat(system="You are a nice assistant")

session = Chat([
   {"role": "system", "content": "You are a nice assistant"}
])
```

This system prompt can always be updated later on:

```python 
session = Chat(system="You are a nice assistant")
session.system = "You are an evil assistant"
```

We can add user and assistant messages to the session:

```python 
from prompts import Message

chat.add_user_message("Hi!")
chat.add_assistant_message("Hi!")
```

Adding two successive `user` or `assistant` messages will raise an exception. To prevent user errors we will create two types `User` and `Assistant` to replace `Message("user"` and `Message("assistant"`:

```python
chat += User("hi!")
chat += Assistant("hi!")
```

We can also easily "fork" discussions:

```python
chat = Chat()
...
new_chat = chat + Message("user", "Mew")
```

And finally we can render the conversation using `model`'s chat template with

```python 
chat.render("model")
```

Note that before rendering, the `render` method calls an internal `filter` method that can trim the discussion (pass through by default). To implement a memory-less agent you would subclass `Chat` this way:

```python
class MarkovianChat(Chat):

    def filter(self):
        return self.history[-1]
```

For vision model we can re-use the `Vision` objects we defined in the new version of Outlines:

```python
chat += user(Vision("prompt", image))
```

## TODO

- [ ] Tests
- [ ] Add `user` and `assistant` aliases
- [ ] Templated discussion

### CODE REVIEW

**Comment by rlouf** [comment]
I will expand on this later, but I think the following user interface os much cleaner:

```python
from prompts import Chat, roles

chat = Chat()
chat.append(roles.system())
chat.append(roles.user())
chat += roles.user()
chat += roles.assistant()
chat.render("mistralai/…")
```

**Comment by just-cameron** [comment]
Jan does this over in the Julia world with PromptingTools.jl, and it's extremely ergonomic. PromptingTools has one of the best interfaces I've seen, so big fan of this.

https://github.com/svilupp/PromptingTools.jl

**Comment by willkurt** [comment]
It's useful to compare this to `AutoTokenizer.apply_chat_template` which currently just takes a list of dicts, but accomplishes the same thing without needing to learn a new class.

The biggest issue I probably have with `apply_chat_template` is that each model has different rules. So for example your point about: "Adding two successive user or assistant messages will raise an exception." only applies to Mistral as far as I know (or at least it doesn't apply to every instruct setup). Phi-3 seems to ignore system prompts (but I can only tell this by printing the prompt out), etc.

I do see some advantages in this proposed approach, namely:

- I can never remember the roles, so having them be properties rather than strings is nice.
   -  Especially good if some roles aren't supported (or in the possible case of additional roles)
- Certainly is less verbose than manually typing dicts

Additionally I do like the template example you have in chat. However it seems to me that templating and constructing prompts themselves should be independent from managing the chat/instruct interface?

Now if the `Chat` class was initialized with the model (rather than render) we could enforce these rules when messages are appended (as proposed in that example) which is much nicer than only getting that feed back when you pass in all the messages. Especially in a notebook environment it would be very helpful to have something like:

```
chat = Chat('phi-3...')
chat.append(roles.system())
>> Sorry, Phi-3 does not allow system prompts, this will be ignored.

chat = Chat('mistral-7b...')
chat.append(roles.user())
chat.append(roles.user())
>>> Sorry, Mistral requires the roles to alternate between 'assistant' and 'user'
```

Of course this diminishes the generality of the Chat as well. Ideally I'm assuming we would like an instance of `Chat` to be rendered however we like.

Basically the tl;dr here is we should make sure that the value add is pretty strong to justify the user switching from a very well known and well supported data structure (list of dicts) to our custom class.

**Comment by rlouf** [comment]
Just to give some context, the interface was designed so rendering can be done automatically in `outlines`:

```python 
from prompts import Chat, user
from outlines import models

openai = models.openai("gpt-4o")

chat = Chat()
chat.system = "You are an AI"
chat += user("Hi!")

result = openai(chat)
```

Where

```python 
chat.render("openai") 
```

is called by the `openai` function.
